lib/min_heap.rb

```diff
@@ -44,28 +50,65 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1)
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return @store.empty?
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(log n)
+  # Space complexity: O(log n) - because recursive calls
   def heap_up(index)
-    
+    return if index.nil? || @store[index].nil?
+
+    parent_index = index/2
+    parent_index -= 1 if index.even?
+
+    heap_up_helper(index, parent_index)
   end
+  
+  def heap_up_helper(index, parent_index)
+    return if @store[index].key > @store[parent_index].key
+
+    swap(index, parent_index)
+    return if parent_index == 0
+
+    index = parent_index
+    parent_index = index/2
+    parent_index -= 1 if index.even?
 
+    heap_up_helper(index, parent_index)
+  end
+  
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    return if index.nil? || @store[index].nil?
+  
+    heap_down_helper(index)
+  end
+   
+  def heap_down_helper(index)
+    left_child = index * 2 + 1
+    right_child = index * 2 + 2
+
+    return if @store[left_child].nil? && @store[right_child].nil?
+
+    if @store[left_child] && @store[left_child].key < @store[index].key
+      swap(index, left_child)
+      heap_down_helper(left_child)
+    end
+    
+    if @store[right_child] && @store[right_child].key < @store[index].key
+      swap(index, right_child)
+      heap_down_helper(right_child)
+    end
```

Comment:
Doing it this way will work, but might lead to extra swaps if the right child is initially less than the left.